### PR TITLE
Fix/v2 split placement to non-container node

### PR DESCRIPTION
**Pull request by carpawell**

### CODE REVIEW

**Comment by carpawell** [comment]
Run is [here](https://github.com/nspcc-dev/neofs-node/actions/runs/8481755114) (was not finished at the comment creation time but should be ok).

Why `PUT` == `object not found`: https://github.com/nspcc-dev/neofs-node/issues/2744.

**Review by cthulhu-rider — APPROVED** [review]

File: pkg/services/object/put/distributed.go [hunk 1]
```diff
@@ -134,15 +135,28 @@ func (t *distributedTarget) Close() (oid.ID, error) {
...
 
-	if len(t.obj.Children()) > 0 || t.obj.Type() == objectSDK.TypeLink {
-		// enabling extra broadcast for linking objects
-		t.traversal.extraBroadcastEnabled = true
+	// v2 split link object validation is an expensive routine and useless if
+	// the node does not belong to the container, since another node is responsible
+	// for the link validation and may decline it, does not matter what this node
+	// thinks about it
+	if t.obj.Type() != objectSDK.TypeLink || t.traverser.KeyInContainer(t.isLocalKey) {
+		if t.objMeta, err = t.fmt.ValidateContent(t.obj); err != nil {
```
Comment (roman-khimov) [inline pkg/services/object/put/distributed.go:157]: I think some syntactic check can still be made here, although it doesn't seem to be critical.

File: pkg/services/object_manager/placement/traverser.go [hunk 2]
```diff
@@ -156,6 +156,27 @@ func (x Node) PublicKey() []byte {
 	return x.key
 }
 
+// KeyInContainer passes nodes' public keys from the calculated placement vectors to the
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/services/object_manager/placement/traverser.go:159]: I can call it as `KeyInContainer(isNotLocalKey())`, so some `IterateContainerKeys()` can be even more appropriate.

File: pkg/services/object/put/distributed.go (hunk 1 above)
Comment (carpawell) [reply pkg/services/object/put/distributed.go:157]: you mean to try to parse the link object but not validate? or what "syntactic check"?

File: pkg/services/object/put/distributed.go [hunk 3]
```diff
@@ -243,7 +249,11 @@ loop:
 	}
 
 	// perform additional container broadcast if needed
-	if t.traversal.submitPrimaryPlacementFinish() {
+	if t.traversalParams.submitPrimaryPlacementFinish() {
+		// reset traversal progress
```
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/services/object/put/distributed.go:268]: traverser has `SubmitSuccess` method. it breaks additional broadcasting if not reset this state

File: pkg/services/object_manager/placement/traverser.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/services/object_manager/placement/traverser.go]: i do not like iteration where i do not need iteration (the same way [here](https://github.com/nspcc-dev/neofs-node/pull/2674#discussion_r1542025537)), i just wanted `bool` but there was no key in the `distributedTarget`, no idea why it is useful but did not want to change this part in a little tests fix PR so let it be `IterateContainerKeys`

File: pkg/services/object/put/distributed.go [hunk 4]
```diff
@@ -20,7 +20,8 @@ type preparedObjectTarget interface {
 }
 
 type distributedTarget struct {
-	traversal traversal
+	traversalParams traversal
```
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/services/object/put/distributed.go]: i was not sure i can understand what `traversal` is while there is `traverser`, so let it be `traversalState` then

File: pkg/services/object/put/distributed.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/services/object/put/distributed.go:157]: Yes, parse, but don't go fetching/checking things.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**